R/cnd-abort.R

```diff
@@ -248,6 +290,325 @@ signal_abort <- function(cnd, file = NULL) {
+#'   field named `call`. An easy way to do this is by passing a `call`
+#'   argument to [abort()]. See also [local_error_call()].
+#'
+#' @param arg,error_arg An argument name as a string. This argument
+#'   will be mentioned in error messages as the input that is at the
+#'   origin of a problem.
+#' @param call,error_call An expression (as returned by e.g.
+#'   `sys.call()`) representing the context in which the error
+#'   occurred. If non-null, the call is stripped from its arguments to
+#'   keep it simple.
+#'
+#'   Can also be an execution environment of a currently running
+#'   function (as returned by e.g. `parent.frame()`). The
+#'   corresponding call is then retrieved.
+#'
+#'   See also [rlang::local_error_call()].
+#'
+#' @name args_error_context
+NULL
+
+#' Find the caller argument for error messages
+#'
+#' @description
+#'
+#' `caller_arg()` is a variant of `substitute()` or [ensym()] that
+#' returns an argument name as a string.
+#'
+#' - The resulting name should be formatted as argument, e.g. using
+#'   the `.arg` in the cli package.
+#'
+#' - Use `@inheritParams rlang::args_error_context` to document an
+#'   `arg` or `error_arg` argument that takes `error_arg()` as default.
+#'
+#' @param arg An argument name in the current function.
+#' @usage NULL
+#'
+#' @examples
+#' arg_checker <- function(x, arg = caller_arg(x), call = caller_env()) {
+#'   cli::cli_abort("{.arg arg} must be a thingy.", call = call)
+#' }
+#'
+#' my_function <- function(my_arg) {
+#'   arg_checker(my_arg)
+#' }
+#'
+#' try(my_function(NULL))
+#' @export
+caller_arg <- function(arg) {
+  arg <- substitute(arg)
+  if (!is_symbol(arg)) {
+    abort(sprintf(
+      "%s must be an argument name.",
+      format_arg("arg")
+    ))
+  }
+
+  expr <- do.call(substitute, list(arg), envir = caller_env())
+  if (!is_symbol(expr)) {
+    abort(sprintf(
+      "%s must be an argument name.",
+      format_arg(as_string(arg))
+    ))
+  }
+
+  as_string(expr)
+}
+
+#' Validate and format a function call for use in error messages
+#'
+#' @description
+#'
+#' - `error_call()` creates a function call ready to be used as `call`
+#'   field of error conditions. This field is displayed by [stop()]
+#'   and [abort()] to give context to an error message.
+#'
+#' - `format_error_call()` passes its input to `error_call()` and
+#'   formats the result as code. Use this function if you are
+#'   generating the "in" part of an error message from a stack frame
+#'   call.
+#'
+#' @param call,error_call An expression (as returned by e.g.
+#'   `sys.call()`) representing the context in which the error
+#'   occurred. If non-null, the call is stripped from its arguments to
+#'   keep it simple.
+#'
+#'   Can also be an execution environment of a currently running
+#'   function (as returned by e.g. `parent.frame()`). The
+#'   corresponding call is then retrieved.
+#'
+#'   See also [rlang::local_error_call()].
```

Comment:
Maybe it's worth documenting these together? I think it might be useful to have a couple of examples here to illustrate the basic ideas of call look up.

Reply:
I've just tried this, it doesn't work very well because `local_error_call()` and `error_call()`/`format_error_call()` have different audiences. The former is for end users of `abort()` whereas the latter two are for environments that display error messages from error conditions, such as testthat. So instead I added another link to `local_error_call()` in the description.